fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #80477
* #80418
* __->__ #80416

`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

bdhirsh committed:
fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh commented on torchgen/gen_functionalization_type.py:
```diff
@@ -363,7 +363,17 @@ def emit_view_functionalization_body(
+            c10::DispatchKeySet({{
+              c10::DispatchKey::FuncTorchDynamicLayerBackMode,
+              c10::DispatchKey::FuncTorchDynamicLayerFrontMode,
+              c10::DispatchKey::Python
```
FWIW, I originally thought it would be simpler just to "exclude all keys in TLS", so we're guaranteed to hit the (meta) backend. That breaks though because we still need BackendSelect. Instead of hardcoding the valid set (something like "exclude all keys in TLS, but re-enable the default TLS), I opted to just exclude the keys that I know will break things:
- `Python`, because of `ProxyTorchDispatchMode`
- `FuncTorch` layer mode keys, because if we redispatch to one of them, then they can trample on the existing TLS that I set

If anyone has a strong opinion about a different way to set up the TLS I'm all ears :)

ezyang replied:
So, if Python is excluded, that means that if a Meta implementation is written from Python we will not see it, is that right?

ezyang replied:
Maybe you should remind me why we need to do a full dispatch here, instead of just using our torchgen powers to find the true underlying native:: kernel and calling it directly haha.

bdhirsh replied:
> Maybe you should remind me why we need to do a full dispatch here, instead of just using our torchgen powers to find the true underlying native:: kernel and calling it directly haha.

We could call the underlying `native::` kernels, but many of them are composite, so we end up with the same problem - they'll redispatch to other operators and look at TLS, so we need to zero it out.

> So, if Python is excluded, that means that if a Meta implementation is written from Python we will not see it, is that right?

Hmmm I don't think so? the registration-from-python-logic registers those functions directly to the `Meta` dispatch key, IIRC. So the Python key is still reserved only for python subclasses / python modes.

albanD replied:
can't we call into `at::meta::func()` here?

albanD replied:
> We can't conveniently call at::meta::func(), since some ops are CompositeExplicitAutograd.

Ho the alias keys also get namespace? Not sure why but I was expecting to have namespaces only for keys and would get the functions there.

In any case, with the increase of dynamic addition to the dispatcher (via python registration) I guess we want to discourage using these namespaces as much as possible.

bdhirsh replied:
> Ho the alias keys also get namespace? 

For better or for worse, today yep :p I guess we could figure out how to change this at some point, it might make LTC codegen a bit cleaner too

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

albanD commented on torchgen/gen_functionalization_type.py:
```diff
@@ -391,12 +401,25 @@ def emit_view_functionalization_body(
+        // Specifically, we want to prevent functionalization and Python modes from running.
+        c10::impl::ExcludeDispatchKeyGuard guard(
+            c10::functorch_transforms_ks |
+            c10::DispatchKeySet({{
```
nit: factor the list of keys to skip out to reduce duplication and prevent them from going out of sync?

albanD commented on test/test_functionalization.py:
```diff
@@ -124,15 +132,19 @@ def f(x):
-$3 = torch._ops.aten.view_copy.default($2, [4, 2])
-$4 = torch._ops.aten.mul.Tensor($3, $3)""")
+        self.assertExpectedInline(logs, """\
+
```
Why all the empty lines?

albanD replied:
Oky!

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot rebase

pytorchmergebot committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]

bdhirsh committed:
Update on "fix functionalization regression introduced by ProxyTorchDispatchMode, migrate testing to make_fx"


`ProxyTorchDispatchMode` was added recently as part of `make_fx`, which was secretly causing the meta tensor calls used inside of functionalization to get baked into the graph. It also wasn't caught because the functionalization tests in core don't use `make_fx`, and the tests in functorch aren't as comprehensive.

Now that `make_fx` is in core, I also ported the functionalization test infra over to use it, which would have caught the regression. This also makes the tests cleaner, since mode-based tracing lets us pick up factory functions in the trace output.

[ghstack-poisoned]
